Override service.istio.io/canonical-name label with empty value for acme solver pod

nak3 opened the pull request:
Istio 1.5 mesh always fails with TestAutoTLS/HTTP01.

https://testgrid.knative.dev/serving#istio-1.5-mesh&include-filter-by-regex=test/e2e/autotls.TestAutoTLS/HTTP01

The reason is that the certname route-[UID]-[tag checksum] is too long
and the label in pod created by Challenge breaches the length
limit. Please see below Challenge object.

To fix, this patch overrides service.istio.io/canonical-name label with empty value.

```
>>> challenges.acme.cert-manager.io (1 objects)
>>> Listing
NAMESPACE       NAME                                                                         STATE     DOMAIN                                                   AGE
serving-tests   route-c5834dab-0e46-4379-932c-987dd1bfe2d9-344131649-3671928679-1528632808   pending   http01.serving-tests.3535.knative-boskos-48.kn-e2e.dev   10m

>>> Details
apiVersion: v1
items:
- apiVersion: acme.cert-manager.io/v1alpha2
  kind: Challenge
  metadata:
    creationTimestamp: "2020-06-21T10:39:39Z"
    finalizers:
    - finalizer.acme.cert-manager.io
    generation: 1
    name: route-c5834dab-0e46-4379-932c-987dd1bfe2d9-344131649-3671928679-1528632808
    ... snip ...
  status:
    presented: false
    processing: true
    reason: 'Pod "cm-acme-http-solver-nbmdr" is invalid: metadata.labels: Invalid
      value: "route-c5834dab-0e46-4379-932c-987dd1bfe2d9-344131649-3671928679-1528632808":
      must be no more than 63 characters'
```

/lint

nak3 commented:
/test pull-knative-serving-istio-1.5-mesh

JRBANCEL commented:
/assign @ZhiminXiang

nak3 commented:
First try passed TestAutoTLS/HTTP01. One more time with adding dump_cluster_state to see the object info.

/test pull-knative-serving-istio-1.5-mesh

nak3 commented:
First try passed TestAutoTLS/HTTP01. One more time with adding dump_cluster_state to see the object info.

/test pull-knative-serving-istio-1.5-mesh

vagababov commented on pkg/reconciler/route/resources/names/names.go:
```diff
@@ -38,5 +41,5 @@ func Ingress(route kmeta.Accessor) string {
 // child resource for the given Route.
 func Certificate(route kmeta.Accessor) string {
-	return "route-" + string(route.GetUID())
+	return fmt.Sprintf("route-%d", adler32.Checksum([]byte(route.GetUID())))
```
```suggestion
	return "route-"+strconv.Itoa(int(adler32.Checksum([]byte(route.GetUID()))))
```
the cast might be redundant :)

JRBANCEL replied:
`adler32.Checksum` returns a `uint32` and `strconv.Itoa` wants an `int`.
https://play.golang.org/p/YzWY7ujL3dO

vagababov replied:
So I thought, so we need the cast 📦

ZhiminXiang commented:
Thanks for your PR, Kenjiro.

I am a bit worried that this PR could be a breaking change as it will cause re-creation of all existing Certificates. I am thinking about if there could be a backward compatible way to handle this.

Also, I found that the HTTP01 test passes in Istio 1.5-no-mesh and Istio 1.4, while the longer UID seems to be related to k8s, but unrelated to Istio. Just curious if you have a chance to figure out what caused the longer UID.

nak3 commented:
/test pull-knative-serving-istio-1.5-mesh

nak3 commented:
I understand and agree about the concern. Then, we have another solution.

The culprit is a canonical service label introduced by Istio v1.5 (https://github.com/istio/istio/commit/5f187ac78dabf73de1635a6d03c9501210148ea1). 
Istio (not our istio-webhook) adds the challenge's name (`route-UID-[tag checksum]`) to the value of the labels `service.istio.io/canonical-name`. And the value seems to be assigned because pod's owner is Challenge resource.

So we can override the label in ClusterIssuer with empty value. The canonical label is just used in UI(e.g. Topology view) so it should be no problem. (At least our CI has no issue.)

I reverted the initial change and updated the ClusterIssuer.

nak3 commented:
> The culprit is a canonical service label introduced by Istio v1.5 (istio/istio@5f187ac).
> Istio (not our istio-webhook) adds the challenge's name (route-UID-[tag checksum]) to the value of the labels service.istio.io/canonical-name. And the value seems to be assigned because pod's owner is Challenge resource.


More specifically, if I created following certs:

```
apiVersion: cert-manager.io/v1alpha2
kind: Certificate
metadata:
  name: test-http01
  namespace: serving-tests
spec:
  secretName: test-http01
  duration: 2160h
  renewBefore: 480h
  organization:
  - prodibiv2
  dnsNames:
  - knative.example.com
  issuerRef:
    name: http01-issuer
    kind: ClusterIssuer
```

Then, cert-manager creates the following challenge:


```
$ kubectl get challenge -n serving-tests test-http01-1530182310-2267070445-1750222759 
NAME                                           STATE     DOMAIN                AGE
test-http01-1530182310-2267070445-1750222759   pending   knative.example.com   62s
```

And the following pod with label will be created by cert-manager.

```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    sidecar.istio.io/inject: "true"
    sidecar.istio.io/status: '{"version":"fca84600f9d5ec316cf1cf577da902f38bac258ab0fd595ee208ec0203dc0c6d","initContainers":["istio-init"],"containers":["istio-proxy"],"volumes":["istio-envoy","podinfo","istiod-ca-cert"],"imagePullSecrets":null}'
  creationTimestamp: "2020-06-29T11:15:56Z"
  generateName: cm-acme-http-solver-
  labels:
    acme.cert-manager.io/http-domain: "1273628538"
    acme.cert-manager.io/http-token: "801902113"
    acme.cert-manager.io/http01-solver: "true"
    security.istio.io/tlsMode: istio
    service.istio.io/canonical-name: test-http01-1530182310-2267070445-1750222759
    service.istio.io/canonical-revision: latest
  name: cm-acme-http-solver-zgchn
  namespace: serving-tests

 ...

```

ZhiminXiang commented:
/lgtm
/approve

@nak3 thanks a lot for the fix!

ZhiminXiang commented:
/assign @chizhg 
/assign @chaodaiG

chizhg commented:
/approve
